WordPress/gutenberg
# ToggleGroupControl: rewrite backdrop animation with framer motion shared layout animations

## What?


Rewrite `ToggleGroupControl`s backdrop animation using `framer-motion`'s [shared layout animations](https://www.framer.com/motion/layout-animations/#shared-layout-animations).

This PR supersedes https://github.com/WordPress/gutenberg/pull/49822, since in the meantime:
- `framer-motion` got updated to a more recent version (https://github.com/WordPress/gutenberg/pull/49822)
- `ToggleGroupControl` went through some major internal refactoring (https://github.com/WordPress/gutenberg/pull/45123 in particular)


## Why?


The current implementation for the backdrop animation is complex, difficult to read, imperative (based off arbitrary timeouts), and [buggy](https://github.com/WordPress/gutenberg/issues/37506).

## How?



Main changes are:

- deleted the `ToggleGroupControlBackdrop`, and the imperative logic that came with it
- re-implemented the backdrop using `framer-motion`'s shared layout animations:
    - the main way this works is by using the same value for the `layoutId` attribute to tag the components that participate to the same shared layout animation
    - each `ToggleGroupControl` also defines its own "namespace" for the shared layout animations via the `LayoutGroup` component, in order to avoid conflicts between different instances of the component
- cleanup up some styles and removed some unused variables
- memoized some classnames
- tweaked the  `InspectorControl`'s `Slot` and `Fill` components to forward the `MotionContext`
- rewrote the internal controlled/uncontrolled logic, that previously relied on `usePrevious` + `useUpdateEffect`. The refactor meant:
  - using `useControlledValue` for the "as button" version of the component
  - refactoring from `reakit` to `ariakit` for the "as radio" version of the component (since it has support for controlled/uncontrolled logic)
  - adding some extra logic to distinguish between controlled/uncontrolled when an `undefined` value is passed and based on the previous value of the `value` prop (to maintain backwards-compat)

## Testing Instructions





- In storybook, go over all the examples for the `ToggleGroupControl` component and make sure that the backdrop animates as expected
- In the post editor, select a paragraph block, and make sure that the font size picker (rendering a `ToggleGroupControl`) works and animates as expected
- In the site editor, open the global styles sidebar:
  - Select "Typography", then one of the list elements. Make sure that the font size picker (rendering a `ToggleGroupControl`) works and animates as expected
  - Select a heading. Make sure that the font size picker (rendering a `ToggleGroupControl`) works and animates as expected
### Testing Instructions for Keyboard


## Screenshots or screencast 


https://github.com/WordPress/gutenberg/assets/1083581/69b6c703-0a78-4b63-8cfd-fd7020215957
